### ISSUE TITLE
[2.2rc3] Keras validation data doesn't respect cache in MirroredStrategy

### ISSUE THREAD

**Issue by lgeiger** [tensorflow/tensorflow#38655]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.2.0rc3
- Python version: 3.7
- CUDA/cuDNN version: 10.1 / 7.6.5.32
- GPU model and memory: 4 x NVIDIA V100 on GCP

**Describe the current behavior**

When running the code below with cached training and validations datasets in a multi-GPU environment (I am using a GCP VM with 312GB of memory and 4 NVIDIA V100s) with `tf.distribute.MirroredStrategy()` the validation dataset isn't correctly cached and examples are still read from GCS during validation.

The memory usage suggests that the validation dataset is cached, but during the Keras validation loop it looks like that data is still read from GCS instead of from the cache which can be observed by the very high network usage during validation. I would expect no network usage after the first epoch.

In the example below I intentionally use an very large validation set to make this issue very obvious and easy to detect through monitoring network usage. This behaviour can also be observed with other datasets, but the unexpected network access will be less noticible on smaller datsets.

**In which cases can this issue not be observed?**
To narrow down the possible causes for this I found two cases where this issue doesn't exist:

1. When running on a single GPU without `MirroredStrategy` the validation data is correctly read from the cache and after the start of the second epoch no additional network traffic reading from GCS can be observed.

2. When not using a validation dataset at all the network usage is zero after the first epoch so caching of the training set works as expected.

This seems to be a complicated interaction between `tf.data`, `tf.keras` and `tf.distribute`, do you have an idea what could cause this behaviour? Please let me know what additional information I could provide.

**Describe the expected behavior**

Network usage should be zero after the start of the second epoch since both datasets are cached  in memory and no additional reads from GCS should be required.

**Standalone code to reproduce the issue**
```python
import tensorflow as tf
import tensorflow_datasets as tfds


batch_size = 1024
decoders = {"image": tfds.decode.SkipDecoding()}

dataset = tfds.load(
    "imagenet2012:5.0.0",
    decoders=decoders,
    split="validation",
    data_dir="gs://my-data-bucket",
)

val_dataset = tfds.load(
    "imagenet2012:5.0.0",
    decoders=decoders,
    split="train",
    data_dir="gs://my-data-bucket",
)


def _decode_and_center_crop(image_bytes):
    """Crops to center of image with padding then scales image_size."""
    shape = tf.image.extract_jpeg_shape(image_bytes)
    image_height = shape[0]
    image_width = shape[1]
    image_size = 224

    padded_center_crop_size = tf.cast(
        (
            (image_size / (image_size + 32))
            * tf.cast(tf.minimum(image_height, image_width), tf.float32)
        ),
        tf.int32,
    )

    offset_height = ((image_height - padded_center_crop_size) + 1) // 2
    offset_width = ((image_width - padded_center_crop_size) + 1) // 2
    crop_window = tf.stack(
        [offset_height, offset_width, padded_center_crop_size, padded_center_crop_size]
    )
    image = tf.image.decode_and_crop_jpeg(image_bytes, crop_window, channels=3)
    return tf.image.resize(image, [image_size, image_size], method="bicubic")


def preprocessing(data):
    return tf.cast(_decode_and_center_crop(data["image"]), tf.float32), data["label"]


def apply_preprocessing(dataset):
    return (
        dataset.cache()
        .map(preprocessing, num_parallel_calls=tf.data.experimental.AUTOTUNE)
        .batch(batch_size)
        .prefetch(1)
    )


dataset = apply_preprocessing(dataset)
val_dataset = apply_preprocessing(val_dataset)

with tf.distribute.MirroredStrategy().scope():
    model = tf.keras.models.Sequential(
        [
            tf.keras.layers.GlobalMaxPool2D(input_shape=(224, 224, 3)),
            tf.keras.layers.Dense(1000, activation="softmax",),
        ]
    )

    model.compile(
        optimizer="adam",
        loss="sparse_categorical_crossentropy",
        metrics=["accuracy", "sparse_top_k_categorical_accuracy"],
    )

model.fit(
    dataset, epochs=5, validation_data=val_dataset,
)
```

**Other info / logs**

To monitor the network usage over time tools like [`ytop`](https://github.com/cjbassi/ytop/) can be used.

**Comment by guptapriya** [comment]
Thank you for filing, it seems like this is known issue and the team is looking into it.  I can suggest a workaround with tf-nightly if this is blocking you.

**Comment by Saduf2019** [comment]
@lgeiger 
please update as per above comment

**Comment by byronyi** [comment]
> Thank you for filing, it seems like this is known issue and the team is looking into it. I can suggest a workaround with tf-nightly if this is blocking you.

It'd be great to have a workaround, in case the fix cannot make into v2.2.

**Comment by guptapriya** [comment]
Actually @jsimsa just submitted a fix for this yesterday (https://github.com/tensorflow/tensorflow/commit/7ebbab819e736319ec35b48e31f4d62fbad6626b). Assuming that has made it to the nightly - could @lgeiger could you try the nightly and see if this has been fixed for your use case? 
We will try to have this fix in 2.2 as well.

**Comment by lgeiger** [comment]
@guptapriya @jsimsa Thank you for the fast fix 🎉 

I can confirm that this issue doesn't exist on the latest nightly.

Should I open a PR to cherry-pick 7ebbab819e736319ec35b48e31f4d62fbad6626b onto the release branch?

**Comment by mihaimaruseac** [comment]
Has been cherrypicked as #38807

**Comment by lgeiger** [comment]
Thanks for the fast fix 👍

**Comment by alimhanif** [comment]
Hi, I was training my model using Multi Workers Mirrored Strategy and still get this issue.
Time by time, the memory usage is still increasing until I got OOM.

**Comment by lgeiger** [comment]
@alimhanif Could you check if your issue is related to #38617?
